src/services/patternMatcher.ts

```diff
@@ -0,0 +1,813 @@
+        // correspond to "U", "I" and "Element".  If "UIElement" isn't found as an exaxt, prefix.
+        // or substring match, then the character spans will be used to attempt a camel case match.
+        characterSpans: TextSpan[];
+    }
+
+    function createPatternMatch(kind: PatternMatchKind, punctuationStripped: boolean, isCaseSensitive: boolean, camelCaseWeight?: number): PatternMatch {
+        return {
+            kind,
+            punctuationStripped,
+            isCaseSensitive,
+            camelCaseWeight
+        };
+    }
+
+    export function createPatternMatcher(pattern: string): PatternMatcher {
+        // We'll often see the same candidate string many times when searching (For example, when
+        // we see the name of a module that is used everywhere, or the name of an overload).  As 
+        // such, we cache the information we compute about the candidate for the life of this 
+        // pattern matcher so we don't have to compute it multiple times.
+        var stringToWordSpans: Map<TextSpan[]> = {};
+
+        pattern = pattern.trim();
+
+        var fullPatternSegment = createSegment(pattern);
+        var dotSeparatedSegments = pattern.split(".").map(p => createSegment(p.trim()));
+        var invalidPattern = dotSeparatedSegments.length === 0 || forEach(dotSeparatedSegments, segmentIsInvalid);
+
+        return {
+            getMatches,
+            getMatchesForLastSegmentOfPattern, 
+            patternContainsDots: dotSeparatedSegments.length > 1
+        };
+
+        // Quick checks so we can bail out when asked to match a candidate.
+        function skipMatch(candidate: string) {
+            return invalidPattern || !candidate;
+        }
+        
+        function getMatchesForLastSegmentOfPattern(candidate: string): PatternMatch[] {
+            if (skipMatch(candidate)) {
+                return undefined;
+            }
+
+            return matchSegment(candidate, lastOrUndefined(dotSeparatedSegments));
+        }
+
+        function getMatches(candidate: string, dottedContainer: string): PatternMatch[] {
+            if (skipMatch(candidate)) {
+                return undefined;
+            }
+
+            // First, check that the last part of the dot separated pattern matches the name of the
+            // candidate.  If not, then there's no point in proceeding and doing the more
+            // expensive work.
+            var candidateMatch = matchSegment(candidate, lastOrUndefined(dotSeparatedSegments));
+            if (!candidateMatch) {
+                return undefined;
+            }
+
+            dottedContainer = dottedContainer || "";
+            var containerParts = dottedContainer.split(".");
+
+            // -1 because the last part was checked against the name, and only the rest
+            // of the parts are checked against the container.
+            if (dotSeparatedSegments.length - 1 > containerParts.length) {

```

Comment:
Does containerParts always represent a name that is fully qualified? That's the assumption you seem to be relying on.